File: onset_times_events.m
close all 
clear all
%---------------------------------------------------------------------------
% This script takes file(s) the following inputs:
% (i) path to input file(s)
% (ii) format the filenames (.ev2) containing details of events recorded 
% (iii) def. of events (neurologist markings, e.g. 21 and 41, ...),
% (iv) eeg sampling frequency 
% (v) fmri trigger in seconds (if known, optional)
% and outputs struct. array named terms which contains 
% the following:
% (i): number of conditions established in each run
% (ii): string of definition of each event in each run
% e.g. type_1_22_run_3
% (iii): definition of each event (i.e. types each event contains) (array)
% (iv): onsettimes array that (col. 1) lists the times provided in input 
% dataset, (col. 2) onset times in unit of seconds with respec to the first 
% fmri image acquistion
% (iv): onset times of each event in unit of (s) from col. 2 of onsettimes array
% (v): frequency of occurence of each event in each run

% Remarks: 

% *** assume format of input file(s) as follows: 
% 2nd column denotes event types, 
% 3rd column denotes response, and
% 6th column denotes event times in (seconds), but sometimes they are
% expressed as integer offsets instead, which correspond to the number of 
% EEG samples from the beginning of the recording.
% ***

% - directory specified only stores the necessary number of ev.2 file(s)
% - string "runX_ev" must be contained in input filename

% - diff. between onset_times_events_1.m and onset_times_event.m are
% that 
% (i): when onset time (ti) of an event is not found in a certain run, instead of
% setting ti of that regressor as 0 s, this new version simply doesn't even
% create a regressor for that event type in that run. Thus, total number of
% conditions (regressors) in all runs depends on the number of event types
% sucessfully identified in input file(s).
% (ii): less manual labor work is required for running, in that only
% entering number of runs and format of input filename, and the rest will be
% handled.
% (iii): when no event is identified in a run, an empty cell is assigned to 
% the onset times array, a new field "results" is constructued to store
% non-empty entries of all relevant arrays, while keeping the orginal
% (untrimmed) data in separate field "data".

% Updates: 
% (i) March 4, 2021, instead of using eventmat(1,6) (1st row, 6th column)
% as the time when the first func. image was aquired, find the first
% occurrence of such event (type 0 and response 5), and only consider
% event types that happened after such event was found.

% (ii) March 16, 2021, parenthese in event_def_str is removed to avoid
% confusion. Also, allow the user to enter fmri trigger when *.ev2 file(s) 
% contains no such information (lacking type 0 and response 5).

%---------------------------------------------------------------------------
%---------------------------------------------------------------------------
% BEGIN USER INPUT

% enter subject number (str)
subnum = '27';

% enter directory and filename of input file(s)
directname = ['/Users/michaeltang/Downloads/fmri_project/', 'sub', subnum, '_imthres0_exmask'];

% % enter format of input file(s)
evfilename_format = ['Sub', subnum, '*.ev2'];   % %wildcard char is added for searching matched files in directname

% enter event types interested
% e.g. {[1, 11], [12], [13]} denotes event 1 is composed of types 1 and 11, 
% event 2 is composed of type 12, while event 3 is composed of type 13
events = {[1 11], [12], [13]}; 

% enter output directory and filename 
directnameoutput = [directname, '/matrices'];
filenameoutput = ['onset_times_events_sub', subnum, '.mat'];

% enter eeg sampling frequency
sampfreq = 2000;   % Hz

% enter time in seconds of first fmri frame relative to beginning of eeg 
% recording (if known, otherwise leave it empty ([]) and it will be determined
% in the script)
fmri_frame_ti = [];

% enter if output should be written to path (1 = yes, 0 = no)
op_results = 0;

% END USER INPUT
%---------------------------------------------------------------------------
%---------------------------------------------------------------------------

%----------------

% with format of *.ev2 filename entered, use dir to search and 
% get full path of file(s)
evfile_path = dir(fullfile(directname , evfilename_format));   

% each .ev2 file found is given a dim. in struct. evfile_path to store 
% its info., get number of .ev2 files located in directory 
input_num = size(evfile_path, 1);

% print message to terminal listing number of .ev2 files found for this
% subject
sprintf('subject %s, number of .ev2 files found = %d', subnum, input_num)

% access name of each .ev2 file from struct. evfile_path
evfilename = cell(input_num, 1);
for i = 1:input_num
    evfilename{i} = evfile_path(i).name;
end

% initialize array for outputting onset times and relevant info.
terms = {};
for i = 1:input_num
    fn_run = sprintf('run%d', i);
    terms.(fn_run) = onset_times(directname, evfilename{i}, sampfreq, events, fmri_frame_ti);
end

% output struct. array to directory
if op_results == 1
    save(fullfile(directnameoutput,filenameoutput),'terms');
end

%--------------------------------

% define function named onset_times that requires the following inputs:
% (i) directory of event details file, 
% (ii) filename of event details (*.ev2 files), 
% (iii) sampling freq., 
% (iv) array(s) of event types interested, and 
% (v) time in seconds when the first fmri frame was acquired (optional)
% then output the following: 
% (i) number of conditions established in each run
% (ii) strings of event definitions
% (iii) onset times matrix (in terms of integer offset and time)
% (iv) events' intitial times (unit of seconds)
% (v) freq. of each event
% remarks: input arguments of function must be placed in the order defined

function run = onset_times(directname, evfilename, sampfreq, events, fmri_trigger)

    % import data from files to struct
    eventdata = importdata(fullfile(directname, evfilename));   % event data for this run
    
    % get event matrix of each run from struct.
    eventmat = eventdata.data;  
    
    % get number of events provided 
    numevents = length(events);
    
    % create eventdef array, use for loop to assign the i^th variable input
    % as the i^th event (e.g. define types 21 and 41 as event 1, types 12
    % as event 2, and so on ...)
    eventdef = cell(1, numevents);
    for i = 1:numevents
        eventdef{i} = events{i};
    end

    % create empty cell array for storing onset times (both recorded time (col. 1) 
    % and time (s) relative to the first fmri image (col. 2))
    ontimesar = {};
    
    % initialize array evfre for storing freq. of each event 
    evfreq = cell(1, numevents);
    for i = 1:numevents
        evfreq{i} = 0;
    end
    
    % check format of onset times reported in dataset 
    % i.e. are times reported in (seconds) or (integer offset)
    % formatnum = 0 means onset times in dataset are in (seconds)
    % formatnum = 1 means onset times in dataset are in (offset)
    % note: output onset times are all in unit of seconds relative to the 
    % start of the beginning of fmri acquisition
    formatnum = logical(eventmat(1,6)/sampfreq > 1);
   
    % determine fmri_ti, the time at which first fmri frame was aqcuired
    % relative to the start of eeg recordings, by checking if the 5th input
    % arg. is an empty matrix. If so, locate the first occurrence of type 0 
    % and response 5, otherwise, assign the 5^th input arg. to the variable.
    if isempty(fmri_trigger)   % if 5th input arg. is empty
        % get time and row number in eventmat at which first func. image was 
        % acquired (fmri_ti), which corresponds to neurological marking 
        % of event type 0 and respone 5
        for i = 1:size(eventmat, 1)   % each row
            if eventmat(i, 2) == 0 && eventmat(i, 3) == 5   % col. 2 = type and col. 3 = response
                fmri_ti = eventmat(i, 6);   % get offset from col. 6 
                fmri_ti_rownum = i;   % get correpsonding row number
                break   % break loop as only first occurrence is interested
            end
        end
        % print fmri_ti and the associated row number in current input file to terminal
        sprintf('In %s, row %d, first fmri frame was aquired at %.3f (offset or seconds)', ...
            evfilename, fmri_ti_rownum, fmri_ti)
    end
    
    % if the 5th input arg. is not empty, assign the 5^th input arg. 
    % to fmri_ti and set assoicated row number as 1, so all rows including
    % the 1st one in input file (.ev2) are scanned in the loop defined below
    if ~(isempty(fmri_trigger)) % if 5th input arg. is not empty
        fmri_ti = fmri_trigger;   % assign the 5^th input arg. to var.
        fmri_ti_rownum = 1;   % set row number
        sprintf('In %s, first fmri frame was set as the 5th input arg. (%.3f)', evfilename, fmri_trigger)
    end
   
    % use for loop to go through each event, for each value in the def. 
    % of the i^th event, check if it is equal to any row in col. 2 of eventmat.
    % Only consider rows after fmri_ti (happened after fmri acquisition
    % had begun)
    % (e.g. def. of event{i} = [1, 11], check which row in col. 2 of
    % eventmat has the same value (1 or 11))
    % if so, extract the corresponding onset time from column 6, 
    % check unit of times reported, and convert the values to unit of
    % seconds if necessary, assign resultant value to ontimesar
    % e.g. ontimesar{3}(5,2) contains the 5^th onset time in (s) of event 3
    
    for i = 1:numevents   % for each event
        for row = fmri_ti_rownum:size(eventmat, 1)   % for each row in eventmat after fmri aquisition began
            for j = 1:length(eventdef{i})   % for each value in def. of event
                if eventmat(row, 2) == eventdef{i}(j)   % if entry in eventmat matches with value in def.
                    evfreq{i} = evfreq{i} + 1;   % increment ev. freq. by 1
                    ontimesar{i}(evfreq{i}, 1) = eventmat(row, 6);   % get onset time from col. 6 in eventmat
                    if formatnum == 0   % check unit of onset times
                        ontimesar{i}(evfreq{i}, 2) = (ontimesar{i}(evfreq{i}, 1) - fmri_ti);   % calcu. onset times when unit is in seconds
                    end
                    if formatnum == 1
                        ontimesar{i}(evfreq{i}, 2) = (ontimesar{i}(evfreq{i}, 1) - fmri_ti)/sampfreq;   % calcu. onset times when unit is in offset
                    end
                end
            end
        end
        if evfreq{i} == 0   % if no rows in eventmat matched with current def. of events
            ontimesar{i} = [];   % assign empty matrix
        end
    end
    
    % extract initial times (ti) in unit of seconds of each event type from ontimesar
    evti = cell(1, numevents);
    for i = 1:numevents
        if isempty(ontimesar{i})   % if current cell is empty (no marking found)
            evti{i} = [];
        else   % if current cell is not empty
            evti{i} = ontimesar{i}(:,2);   % from col. 2 in each cell 
        end
    end
    
    % create string describing event details in each run 
    sind = regexp(lower(evfilename), 'run');   % start index of 'r' in event filename
    eind = strfind(lower(evfilename), '_ev');   % index of '_ev' in event filename
    runnum = evfilename(sind+3:eind-1);   % string of run number contained in event filename
    
    % assemble required strings
    eventdefstr = cell(1, numevents);
    for i = 1:numevents
        % create place holder by num. of types in each event
        placestr = repmat('%d_', 1, length(eventdef{i})); 
        % assign formatted string to eventdefstr array
        eventdefstr{i} = sprintf(append('type_', placestr, 'run_%s'), eventdef{i}, runnum);
    end
 
    % empty matrix is assigned to ontimesar if no event type is found in
    % eventmat, only select non-empty cells  
    
    % create matrix named nonemptyind for storing indices of non-zero
    % entries in evfreq
    nonemptyind = [];
    for j = 1:length(evfreq)
        if evfreq{j} ~= 0
            nonemptyind = [nonemptyind j];
        end
    end
    
    % number of conditions established in each run is equal to the number 
    % of sets of non-empty event types identified, get length of nonemptyid 
    % to assign value to var. num_cond
    num_cond = length(nonemptyind);
     
    % empty matrix is assigned to ontimesar if no event type is found in eventmat
    % only select non-empty cells (i.e. event with non-zero freq. recorded)
    % assign NON-EMPTY values to field "results" under struct. "run"   
    run.results.num_cond = num_cond;   % number of conditions estaliblished in each run
    run.results.event_def_str = eventdefstr(nonemptyind);   % string describing def. of each event for spec. run
    run.results.event_def = eventdef(nonemptyind);    % defition of each event for spec. run
    run.results.onsettimes = ontimesar(nonemptyind);   % onset times recorded in dataset provided
    run.results.evti = evti(nonemptyind);   % initial time (seconds) of each event in each run
    
     % output quantities calcu. (including EMPTY values) to field "data" 
     % under struct. "run"
    run.data.event_def_str = eventdefstr;   % string describing def. of each event for spec. run
    run.data.event_def = eventdef;   % defition of each event for spec. run
    run.data.onsettimes = ontimesar;   % onset times recorded in dataset provided
    run.data.evti = evti;   % initial time (seconds) of each event in (s)
    run.data.evfreq = evfreq;   % freq. of each event 
        
    
end   % end function defined
%--------------------------------

%---------------------------------------------------------------------------

File: onset_times_events.py
globals().clear()

import re
#import pandas as pd
from pandas import read_csv
import numpy as np
from prelim_setup import get_path

#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------

# BEGIN USER INPUT

# enter subject number (str)
subnum = '14'

# enter directory and filename of input file(s)
directname = '/work/levan_lab/mtang/fmri_project/' + 'sub' + subnum
# directname = ['/Users/michaeltang/Downloads/fmri_project/', 'sub', subnum, '_imthres0_exmask'];

# enter formats of input file(s)
evfilename_format = ['Sub' + subnum + '.*\.ev2', '^Run\d+_.*\.txt']

# enter output directory and filename 
directname_output = directname + '/py_matrices/'
filename_output = 'onset_times_events_sub' + subnum + '.npy'

# enter eeg sampling frequency in Hz
sampfreq = 2000

# enter time in seconds of first fmri frame relative to beginning of eeg 
# recording (if known, otherwise leave it empty ([]) and it will be determined 
# in the script)
if subnum == '35':
    fmri_frame_ti = 34.3647
else:
    fmri_frame_ti = [];


# END USER INPUT
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------

# format full path of event types file
ev_type_path = '/work/levan_lab/mtang/fmri_project/event_types.txt'

# load data from event types file into dataframe
ev_type_df = read_csv(ev_type_path)

# find row with matching subject number in ev_type_df
row_matched = ev_type_df.iloc[:, 0] == int(subnum)

# obtain subset of dataframe using row number found, convert subset to list
ev_type_list = ev_type_df[row_matched].iloc[:, -1].tolist()

# remove whie space and split string by ' ' 
ev_type_str = ev_type_list[0].strip().split(' ')

# event type with more than one def. contains the pattern '-', which is used 
# separate each def. check if an entry contains '-', if so, further split string
# into components by '-', append data to ev_type_read list
ev_type_read = []   # intiailize list
for entry in ev_type_str:   # for every entry in list
    if '-' in entry:   # if '-' exists in current entry
        ev_type_read.append(entry.split('-'))   # split string by '-', append to list
    else:   # if '-' doesn't exist in curr. entry, append entry to list
        ev_type_read.append(entry)
        
# assign ev_type_read to events
events = ev_type_read
        
# search for matching filenames in directory specified, use get_path to otain
# full path of event files and its format number
for i in range(len(evfilename_format)):   # every format input
    if get_path(directname, evfilename_format[i]):   # if matching filename is found
        ev_file_path = get_path(directname, evfilename_format[i])   # obtain full path
        ev_file_format_num = i   # assign ev file format number
        break
    
# format of event files (.ev2 or .txt) determines the way event onsets are obtained:

#----------------------------------------------------------
# obtain event onsets by the following if input files are in .ev2 format (ev_file_format_num == 0) 
if ev_file_format_num == 0:
    
    # initialize output dict for curren run
    op_dict = {}
    
    for run_ind in range(len(ev_file_path)):
        
        # get indices of current run number embedded in path
        s_ind = re.search('Run\d.*_', ev_file_path[run_ind], re.IGNORECASE).span()
        
        # get current run number
        runnum = ev_file_path[run_ind][s_ind[0]+3:s_ind[1]-1]
    
        # load data in .ev2 file into numpy array
        event_array = np.loadtxt(ev_file_path[run_ind])


    # check format of onset times reported in dataset 
    # i.e. are times reported in (seconds) or (integer offset)
    # format_bool = TRUE means onset times in dataset are in (offset)
    # format_bool = FALSE means onset times in dataset are in (seconds)
    # note: output onset times are all in unit of seconds relative to the 
    # start of the beginning of fmri acquisition
        format_bool = event_array[0, -1]/sampfreq > 1

    # determine fmri_ti, the time at which first fmri frame was aqcuired
    # relative to the start of eeg recordings by locating the first occurrence of 
    # type 0 (2nd col.) and response 5 (3rd col.)
        if not bool(fmri_frame_ti):   # if no fmri trigger is entered
            for i in range(event_array.shape[0]):   # every row of event array read
               # identify row of first occurence of event type 0 and response 5
               if event_array[i, 1] == 0 and event_array[i, 2] == 5:
                   # obtain the corresponding time from rightmost col.
                   fmri_ti = event_array[i, -1]   
                   break
        else:   # if fmri trigger is provided by user, assign value to fmri_ti
            fmri_ti = fmri_frame_ti
    
# use for loop to go through each event, for each value in the def. 

# (e.g. def. of event{i} = [1, 11], find matching rows in event array,
# extract the corresponding onset time from rightmost col.
# check unit of times reported, and convert the values to unit of
# seconds if necessary, assign resultant value to evti nested list
# e.g.evti[i^th] contains onset times in seconds for event[i^th]

# initialize ind_total list with format given below,
# in i^th entry contains the matching row indices in event array for the i^th 
# event type. If no matching row indices are found, empty list is assigned for
# that event type
        ind_total = []    
        for i in range(len(events)):   # each def. of event type
        #for i in range(1):
            ev = events[i]   # obtain current event type
            ind_found = []
            # event type could contain more than one def., e.g. [1, 11]
            # find row index in event array with matching event type (2nd col.) for each def.
            for j in range(len(ev)):   # every def. of current event type
            # find matching row index for current event def.
                cur_ind_matched = np.where(event_array[:, 1] == int(ev[j]))[0]   # numpy array
                ind_found.extend(cur_ind_matched.tolist())
                #ind_found.append(cur_ind_matched)
                #ind_found = ind_found + list(cur_ind_matched)   # concatenate lists
            # after looping through all def. of current event type,
            # append all row indices found for current event type to ind_total nested list
            ind_total.append(ind_found)   

# with row indices for every event type obtained, get the corresponding onset
# times in seconds
# initialize evti list format given below
# i^th entry contains onset times in seconds for the i^th event type in events list
# emtpy list is assigned when no onset time is found for the event type
        evti = []
        for item in ind_total:   # every entry in ind_total nested list
            # for cases where there are more than one def. for an event type, 
            # convert indices from list to set, then get the union of the sets to obtain
            # unique row indices, and then sort those indices by ascending order
            ind_req = sorted((set(item)))       
            # check unit of times reported, and calculate onset times with respect to the 
            # start of fmri acquisiton, conert numpy array to list, and append list to 
            # evti nested list
            if ~format_bool:
                evti.append((event_array[ind_req, -1] - fmri_ti).tolist())
            else:
                evti.append(((event_array[ind_req, -1] - fmri_ti)/sampfreq).tolist())
        
            # get frequency of occurence for every event type
            evfreq = [len(item) for item in evti]
            
        # create key for current run
        key_name_cur = 'run' + runnum
        
    # assign event def., event onsets in seconds (evti), and their freq. (evfreq) 
    # to output dict. under current key 'results'
        op_dict[key_name_cur] = {}
        op_dict[key_name_cur]['results'] = {}
        op_dict[key_name_cur]['results']['event_def'] = events
        op_dict[key_name_cur]['results']['evti'] = evti
        op_dict[key_name_cur]['results']['evfreq'] = evfreq
        
#----------------------------------------------------------  
        
# obtain event onsets by the following if input files are in .txt format (ev_file_format_num == 1)     
    
if ev_file_format_num == 1:
    
    # initialize lists with format given below,
    # i^th entry in evti denotes onset times for the run number in i^th entry
    # of runnum, and event def in the i^th entry of ev_def
    # e.g. evti[0] gives onset times for the event type defined in ev_def[i]
    # for the run number defined in runnum[i]
    runnum = []
    ev_def = []
    evti = []
    
    # ger current filename
    for i in range(len(ev_file_path)):
        file_name_cur = re.search('Run\d+_.*\.txt', ev_file_path[i]).group() 
    
        # get run number from cur. file name by first finding the start and end 
        # indices of the search pattern in in cur. file name
        eind = re.search('Run\d+', file_name_cur).span()

        # eind is a tuple with two entries, the first and last index,
        # use the last index to get current run numer
        runnum.append(file_name_cur[3:eind[-1]])
    
        # similarly, get event type contained in cur. file name
        eind_event = re.search('EV\d+', file_name_cur).span()
    
        # use indices obtained above to get current event type
        ev_def.append(file_name_cur[eind_event[0]+2:eind_event[-1]])
        
        # load data from .txt file into numpy arrray
        event_array = np.loadtxt(ev_file_path[i])
        
        # get event onsets from the leftmost column in numpy array loaded
        evti.append(event_array[:, 0].tolist())
        
    # after obtaining all necessary info. from all runs, convert runnum lis to
    # set to get unique entries, and sort the set and convert it backt to list
    unique_runnum = list(sorted(set(runnum)))   # get unique run number
    
    # initialize output dict
    op_dict = {}
    
    # for every unique run number, select corresponding indices in runnum
    # then select the associated event def. and event onsets (evti),
    for item in range(len(unique_runnum)):
        ind_req = [i for i, j in enumerate(runnum) if j == unique_runnum[item]]
        ev_def_cur_run = np.array(ev_def)[ind_req].tolist() 
        evti_cur_run = np.array(evti, dtype = object)[ind_req].tolist()
        evfreq_cur_run = [len(entry) for entry in evti_cur_run]
        
        # after obtaining data for current run, save them in output dict
        fd_name = 'run' + unique_runnum[item]   # name key by current run number
        op_dict[fd_name] = {}   # initialize key
        op_dict[fd_name]['results'] = {}
        op_dict[fd_name]['results']['event_def'] = ev_def_cur_run
        op_dict[fd_name]['results']['evti'] = evti_cur_run
        op_dict[fd_name]['results']['evfreq'] = evfreq_cur_run
        
    
#----------------------------------------------------------  


#sub_list = ['14', '15', '18', '19', '20', '27', '30', '31', '32', '33', '34', \
           #'35', '41']
           
output_path = directname_output + filename_output
    
np.save(output_path, op_dict)            
    
        
        
    
    


